File: zoho/L22-generate_parentheses.py
# https://leetcode.com/problems/generate-parentheses/
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        stack = []
        res = []
        def bt(Open, Close):
            if Open == Close == n:
                res.append("".join(stack))
                return
            if Open < n:
                stack.append("(")
                bt(Open + 1, Close)
                stack.pop()
            if Close < Open:
                stack.append(")")
                bt(Open, Close + 1)
                stack.pop()
        bt(0,0)
        return res

File: zoho/L22-GenerateParentheses.java
// https://leetcode.com/problems/generate-parentheses/
class Solution {
    private static Deque<String> stack = new LinkedList<>();
    private static List<String> res = new ArrayList<>();
    private static void bt(int Open, int Close, int n) {
        if ((Open == Close) && (Open == n) && (Close == n)) {
            res.add(String.join("", stack));
            return;
        }
        if (Open < n) {
            stack.add("(");
            bt(Open + 1, Close, n); 
            stack.removeLast();
        }
        if (Close < Open) {
            stack.add(")");
            bt(Open, Close + 1, n);
            stack.removeLast();
        }
    }
    public List<String> generateParenthesis(int n) {
        bt(0,0,n);
        List<String> res2 = new ArrayList<>(res);
        res.clear();
        return res2;
    }
}
